### PR TITLE
Feat/stacks signer block proposal

### PR DESCRIPTION
**Pull request by jferrant**
### Description

This is a rather large PR...please forgive.

major changes include:
- Complete reorg of the stacker db and stacks clients
- Complete reorg of the signer integration tests for easier adding of tests
- Changing libsigner to take SignerEvents rather than StackerDBEvents as we now need to subscribe to blocks
- Handling the new block proposal events within Stacks signer
- Added a test to handle a validated block (against the RPC endpoint)
- Enabled stackerdb_dkg_sign test as well since I made substantial changes to the test layout to ensure it still works

### Applicable issues

- Closes https://github.com/stacks-network/stacks-core/issues/3928
- Closes https://github.com/stacks-network/stacks-core/issues/3929

### Additional info (benefits, drawbacks, caveats)

Major items left to do that are NOT included in this PR
- State needs to be added to the signers to allow them to filter out already seen blocks requests/block validation responses/block sign requests. See https://github.com/stacks-network/stacks-core/issues/4242
- Block is not properly deserialized during signing rounds to enable verification of a sign requests contents. See https://github.com/stacks-network/stacks-core/issues/3930
- State needs to be added to verify a block contains the expected transactions https://github.com/stacks-network/stacks-core/issues/4243

### Checklist

- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: stacks-signer/src/client/stackerdb.rs
Comment (jcnelson) [inline stacks-signer/src/client/stackerdb.rs:180]: Nit: You probably don't need the clone here; you could just pass `&chunk`.

File: stacks-signer/src/client/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/client/stacks_client.rs:55]: I am not entirely sure. I only really added this per the comment https://github.com/stacks-network/stacks-core/issues/3928#issuecomment-1812947134

I think the main thing is so long as I can verify the signature of the block is valid/belongs to the appropriate miner, it should be fine.

File: stacks-signer/src/client/stackerdb.rs
Comment (jferrant) [reply stacks-signer/src/client/stackerdb.rs:180]: Oh yes definitely here :)

File: stacks-signer/src/client/stackerdb.rs
Comment (jferrant) [reply stacks-signer/src/client/stackerdb.rs:183]: I updated put_chunk to take a reference cause it didn't seem like it actually required ownership so now I can get rid of the clone :D

File: stacks-signer/src/client/mod.rs
Comment (jferrant) [reply stacks-signer/src/client/mod.rs:1]: Done :D I also added it to lib.rs

File: stacks-signer/src/client/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/client/stacks_client.rs:56]: Signature validation happens via the endpoint call. No need for this.

File: Cargo.toml
Comment (jferrant) [inline Cargo.toml:17]: TODO: remove this change. Just temp to access Packet changes I made..

File: libsigner/src/events.rs
Comment (jcnelson) [inline libsigner/src/events.rs:43]: Do you want to use `define_u8_enum!` here?

File: stacks-signer/src/lib.rs
Comment (jcnelson) [inline stacks-signer/src/lib.rs:8]: 2020-2024

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:202]: Will this be done in this PR?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:253]: Presumably, there's also a dependency between validating the block and signing it, right?  What happens if your signer gets a "sign block" request before it has even obtained the block to validate?  Where is this captured in the run loop?

File: stacks-signer/src/lib.rs
Comment (jferrant) [reply stacks-signer/src/lib.rs:8]: Oh right. XD

File: stacks-signer/src/runloop.rs
Comment (jferrant) [reply stacks-signer/src/runloop.rs]: Oh sorry I had actually done this...just hadn't pushed the code LOL

File: stacks-signer/src/runloop.rs
Comment (jferrant) [reply stacks-signer/src/runloop.rs:242]: I will change it around. i.e. as soon as a block arrives from the miner. Trigger a signing round. and then have every signer receiving the sign request pass it to the endpoint to be validated before continuing the signing round. however, this will be heavily dependent on Joey's changes to the sign logic to enable me to deserialize it. I would leave it to a sep PR.

File: stacks-signer/src/lib.rs
Comment (jferrant) [reply stacks-signer/src/lib.rs]: Done

File: stacks-signer/src/runloop.rs
Comment (jferrant) [reply stacks-signer/src/runloop.rs]: Done.

File: libsigner/src/events.rs
Comment (jferrant) [reply libsigner/src/events.rs]: Actually this was dead code from a prior version where I serialized things differently. Deleted!

File: stacks-signer/src/runloop.rs
Comment (kantai) [inline stacks-signer/src/runloop.rs:181]: Should the signer binary track block validation responses directly? This seems like state that will be important for the signer binary to have: that way if a sign request comes in, the signer knows whether or not it has already validated (or rejected) the requested proposal.

Comment (jferrant) [reply stacks-signer/src/runloop.rs:181]: We do have a ticket in plcae to add state to the signers. I do believe we will need it for not only that but to know what transactions we expect to see in a block. I think this should remain a second ticket though considering the size of this change/because I believe we could work on it asyncronously at the same time as the serialization of the block ticket that Joey is assigned to.

Comment (jferrant) [reply stacks-signer/src/runloop.rs:181]: I have made two tickets that track the need for local state for both block requests/responses/sign requests and the need to track transactions to verify a block against

https://github.com/stacks-network/stacks-core/issues/4242
https://github.com/stacks-network/stacks-core/issues/4243

File: stackslib/src/net/api/postblock_proposal.rs
Comment (kantai) [inline stackslib/src/net/api/postblock_proposal.rs:174]: I think it would simplify this function a bit to just have a new type:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct BlockValidateRejectReason {
    pub reason: String,
    pub reason_code: ValidateRejectCode,
}
```

and change the return type of `validate` to `Result<BlockValidateOk, BlockValidateReject>`

with the previous `From<ChainstateError>` implementation.

Then, the caller of `validate()` could do:

```rust
proposal.validate(..).map_err(|reason| BlockValidateReject { block: proposal.block.clone(), reason_code: reason.reason_code, reason: reason.reason })
```

That way, this `validate()` function doesn't have to repeatedly perform that step.

File: stacks-signer/src/runloop.rs
Comment (kantai) [reply stacks-signer/src/runloop.rs:181]: Sounds good to me

File: libsigner/src/events.rs
Comment (kantai) [inline libsigner/src/events.rs:227]: I think you should respond `200` as soon as you've read body into the string. The `stacks-node` will stall progress until it gets a 200 response, and if you exit this function early due to error handling, the `stacks-node` may end up stalling entirely.

File: libsigner/src/events.rs
Comment (kantai) [inline libsigner/src/events.rs:250]: Same thing here -- as soon as the body is read into the string, return 200.

File: testnet/stacks-node/src/tests/signer.rs
Comment (kantai) [inline testnet/stacks-node/src/tests/signer.rs:484]: Should this loop and timeout, e.g.,

```rust
let t_start = Instant::now();
while test_observer::get_proposal_responses().is_empty() {
     assert!(t_start.elapsed() < TIMEOUT, "Timed out while waiting for block proposal event");
     thread::sleep(Duration::from_secs(1));
}
```

Otherwise, the test might have a race condition.

File: testnet/stacks-node/src/tests/signer.rs
Comment (kantai) [inline testnet/stacks-node/src/tests/signer.rs:404]: Can you add a rustdoc comment to this test method that just describes at a high level, what behavior this function is testing and what it does to do so?

**Review by kantai — COMMENTED** [review]
This looks great to me, just had a few comments

File: libsigner/src/events.rs
Comment (jferrant) [reply libsigner/src/events.rs:227]: Done!

File: libsigner/src/events.rs
Comment (jferrant) [reply libsigner/src/events.rs:250]: Done!

File: testnet/stacks-node/src/tests/signer.rs
Comment (jferrant) [reply testnet/stacks-node/src/tests/signer.rs:484]: Done! :)

File: testnet/stacks-node/src/tests/signer.rs
Comment (jferrant) [reply testnet/stacks-node/src/tests/signer.rs:404]: Done!

File: stackslib/src/net/api/postblock_proposal.rs
Comment (jferrant) [reply stackslib/src/net/api/postblock_proposal.rs]: Done!

**Review by kantai — APPROVED** [review]
LGTM!

File: stacks-signer/src/client/stackerdb.rs
Comment (jcnelson) [inline stacks-signer/src/client/stackerdb.rs:1]: Copyright comment block please

File: stacks-signer/src/client/stackerdb.rs
Comment (jcnelson) [inline stacks-signer/src/client/stackerdb.rs:3]: Isn't this `stackslib::` now, instead of `blockstack_lib`?

File: stacks-signer/src/client/stackerdb.rs
Comment (jcnelson) [inline stacks-signer/src/client/stackerdb.rs:83]: Commented-out code

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [reply stacks-signer/src/runloop.rs:242]: Can you add a ticket for this before we merge this?

File: stacks-signer/src/runloop.rs
Comment (jcnelson) [inline stacks-signer/src/runloop.rs:292]: Please don't commit commented-out code

File: stackslib/src/net/api/poststackerdbchunk.rs
Comment (jcnelson) [inline stackslib/src/net/api/poststackerdbchunk.rs:57]: Can you instead put this into `stackslib/src/chainstate/stacks/events.rs`?

**Review by jcnelson — APPROVED** [review]
Just a couple comments. Thanks!

File: stacks-signer/src/client/stackerdb.rs
Comment (jferrant) [reply stacks-signer/src/client/stackerdb.rs:3]: I don't find it that way. as far as I am aware its still blockstack_lib. Can find no instance of stackslib::* throughout the code base.

File: stackslib/src/net/api/poststackerdbchunk.rs
Comment (jferrant) [reply stackslib/src/net/api/poststackerdbchunk.rs]: Done :)

File: stacks-signer/src/runloop.rs
Comment (jferrant) [reply stacks-signer/src/runloop.rs]: Done.

File: stacks-signer/src/client/stackerdb.rs
Comment (jferrant) [reply stacks-signer/src/client/stackerdb.rs]: Done :)

File: stacks-signer/src/client/stackerdb.rs
Comment (jferrant) [reply stacks-signer/src/client/stackerdb.rs:16]: Done. Added to ever file in stacks-signer now.

**Pull request merged**
